The quirks attribute is not available on older kernels.

--- a/drivers/net/wireless/libertas/if_sdio.c
+++ b/drivers/net/wireless/libertas/if_sdio.c
@@ -1010,6 +1010,7 @@ static int if_sdio_probe(struct sdio_fun
 	if (ret)
 		goto disable;
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,32))
 	/* For 1-bit transfers to the 8686 model, we need to enable the
 	 * interrupt flag in the CCCR register. Set the MMC_QUIRK_LENIENT_FN0
 	 * bit to allow access to non-vendor registers. */
@@ -1028,6 +1029,7 @@ static int if_sdio_probe(struct sdio_fun
 		if (ret)
 			goto release_int;
 	}
+#endif
 
 	card->ioport = sdio_readb(func, IF_SDIO_IOPORT, &ret);
 	if (ret)
