--- a/drivers/staging/ath6kl/os/linux/ar6000_drv.c
+++ b/drivers/staging/ath6kl/os/linux/ar6000_drv.c
@@ -317,12 +317,20 @@ static void ar6000_refill_amsdu_rxbufs(s
 static void ar6000_cleanup_amsdu_rxbufs(struct ar6_softc *ar);
 
 static ssize_t
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(2,6,34))
 ar6000_sysfs_bmi_read(struct file *fp, struct kobject *kobj,
+#else
+ar6000_sysfs_bmi_read(struct kobject *kobj,
+#endif
                       struct bin_attribute *bin_attr,
                       char *buf, loff_t pos, size_t count);
 
 static ssize_t
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(2,6,34))
 ar6000_sysfs_bmi_write(struct file *fp, struct kobject *kobj,
+#else
+ar6000_sysfs_bmi_write(struct kobject *kobj,
+#endif
                        struct bin_attribute *bin_attr,
                        char *buf, loff_t pos, size_t count);
 
@@ -755,7 +763,11 @@ static struct bin_attribute bmi_attr = {
 };
 
 static ssize_t
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(2,6,34))
 ar6000_sysfs_bmi_read(struct file *fp, struct kobject *kobj,
+#else
+ar6000_sysfs_bmi_read(struct kobject *kobj,
+#endif
                       struct bin_attribute *bin_attr,
                       char *buf, loff_t pos, size_t count)
 {
@@ -782,7 +794,11 @@ ar6000_sysfs_bmi_read(struct file *fp, s
 }
 
 static ssize_t
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(2,6,34))
 ar6000_sysfs_bmi_write(struct file *fp, struct kobject *kobj,
+#else
+ar6000_sysfs_bmi_write(struct kobject *kobj,
+#endif
                        struct bin_attribute *bin_attr,
                        char *buf, loff_t pos, size_t count)
 {
