--- a/drivers/staging/brcm80211/Makefile
+++ b/drivers/staging/brcm80211/Makefile
@@ -24,10 +24,10 @@
 	-DDBAND					\
 	-DBCMDMA32				\
 	-DBCMNVRAMR				\
-	-Idrivers/staging/brcm80211/sys		\
-	-Idrivers/staging/brcm80211/phy		\
-	-Idrivers/staging/brcm80211/util	\
-	-Idrivers/staging/brcm80211/include
+	-I$(obj)/sys		\
+	-I$(obj)/phy		\
+	-I$(obj)/util	\
+	-I$(obj)/include
 
 PCI_CFLAGS := -DWLC_LOW
 
