--- a/drivers/net/wireless/ath/main.c
+++ b/drivers/net/wireless/ath/main.c
@@ -60,16 +60,23 @@ EXPORT_SYMBOL(ath_rxbuf_alloc);
 int ath_printk(const char *level, struct ath_common *common,
 	       const char *fmt, ...)
 {
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,36))
 	struct va_format vaf;
+#endif
 	va_list args;
 	int rtn;
 
 	va_start(args, fmt);
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,36))
 	vaf.fmt = fmt;
 	vaf.va = &args;
 
 	rtn = printk("%sath: %pV", level, &vaf);
+#else
+	printk("%sath: ", level);
+	rtn = vprintk(fmt, args);
+#endif
 
 	va_end(args);
 
