--- a/drivers/net/wireless/iwlegacy/iwl-led.c
+++ b/drivers/net/wireless/iwlegacy/iwl-led.c
@@ -167,7 +167,9 @@ void iwl_legacy_leds_init(struct iwl_pri
 				   wiphy_name(priv->hw->wiphy));
 	priv->led.brightness_set = iwl_legacy_led_brightness_set;
 	priv->led.blink_set = iwl_legacy_led_blink_set;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,30)
 	priv->led.max_brightness = 1;
+#endif
 
 	switch (mode) {
 	case IWL_LED_DEFAULT:
--- a/drivers/net/wireless/iwlwifi/iwl-led.c
+++ b/drivers/net/wireless/iwlwifi/iwl-led.c
@@ -191,7 +191,9 @@ void iwl_leds_init(struct iwl_priv *priv
 				   wiphy_name(priv->hw->wiphy));
 	priv->led.brightness_set = iwl_led_brightness_set;
 	priv->led.blink_set = iwl_led_blink_set;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,30)
 	priv->led.max_brightness = 1;
+#endif
 
 	switch (mode) {
 	case IWL_LED_DEFAULT:
