This backports:

commit 502a2ffd7376ae27cfde6172257db0ff9d8cfec2
Author: stephen hemminger <shemminger@vyatta.com>
Date:   Wed Mar 17 20:31:13 2010 +0000

    ipv6: convert idev_list to list macros

--- a/drivers/net/wireless/iwlwifi/mvm/d3.c
+++ b/drivers/net/wireless/iwlwifi/mvm/d3.c
@@ -99,7 +99,11 @@ void iwl_mvm_ipv6_addr_change(struct iee
 	int idx = 0;
 
 	read_lock_bh(&idev->lock);
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,35))
 	list_for_each_entry(ifa, &idev->addr_list, if_list) {
+#else
+	for (ifa = idev->addr_list; ifa; ifa = ifa->if_next) {
+#endif
 		mvmvif->target_ipv6_addrs[idx] = ifa->addr;
 		idx++;
 		if (idx >= IWL_PROTO_OFFLOAD_NUM_IPV6_ADDRS)
